### PR TITLE
[Feature](bangc-ops): add new operator bbox_overlaps.

### PR DESCRIPTION
**Pull request by mahxn0**
## bbox_overlaps算子测试报告

**_本测试报告模板是希望能帮助算子开发者在完成算子开发后进行有效充分的自检，开发出功能、性能都满足要求的高质量算子。_**

### 1. 修改描述

添加算子描述

- 影响范围/算子：bbox_overlaps
- 影响版本/分支：master

### 1.1 精度验收标准

根据算子需求给出算子分类及其对应的精度验收标准

算子采用静态阈值标准：diffs=[diff1, diff2], diff1<=3e-3 && diff2 <= 3e-3

详细见 [MLU-OPS 精度验收标准](./MLU-OPS-Accuracy-Acceptance-Standard.md)

### 1.2 算子方案CHECKLIST

| 序号 | 需求     | 需求详情                           |
| ---- | -------- | ---------------------------------- |
| 1    | 支持硬件 | MLU370<br>MLU590                   |
| 2    | job类型  | U1 <br> U4                         |
| 3    | layout   | 支持NHWC 、NCHW、ARRAY等layout     |
| 4    | 多维     | 否                                 |
| 5    | 0元素    | 仅bbox1/bbox2输入的第一维支持0元素 |
| 6    | 数据类型 | half / float                       |
| 7    | 规模限制 | 无                                 |

### 1.3 新特性测试

- [x] 数据类型测试
- [ ] 多维张量测试
- [x] Layout 测试
- [x] 不同规模 / 整数余数端段 / 对齐不对齐测试
- [x] 零维张量测试/ 0 元素测试
- [x] 稳定性测试
- [x] 多平台测试
- [x] gen_case模块测试
- [x] nan / inf测试  
- [x] 内存泄漏检查, 详见[GTest-User-Guide-zh](./GTest-User-Guide-zh.md)
- [x] 代码覆盖率检查，详见[GTest-User-Guide-zh](./GTest-User-Guide-zh.md)
- [x] IO计算效率检查，详见[MLU-OPS性能验收标准](./MLU-OPS-Performance-Acceptance-Standard.md) 

**_关于功能测试，是希望算子开发者能够针对每一个逻辑，每一个分支，每一个判断，进行对应的测例设计，尽量覆盖到所有代码逻辑。_**

### 1.4 参数检查

提交新算子时，给出测试点，并说明测试结果。

| 测试点         | 验收标准 | 测试结果（出错信息）                                         |
| -------------- | -------- | ------------------------------------------------------------ |
| 不符合算子限制 | 正常报错 | 1.mode/offset只能是0 \|\| 1 <br>[2023-1-10 17:15:8] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: The mode must be 0 or 1, but now is 3.<br>[2023-1-10 17:16:30] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: The offset must be 0 or 1,  but now is 2.<br>2.如果 bbox1_desc和bbox2_desc第一维不等于0，那么它们的第二维必须等与4<br>[2023-1-10 17:19:44] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: The Boxes' last dimenstion should be 4 or the first dimension should be 0. But now bbox1's last dimension is 5, bbox1's first dimension is 100.<br>3.ious_desc的第一维必须和bbox1_desc的第一维相等<br>[2023-1-10 17:21:17] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: Whether it is aligned mode or not,ious_desc->dims[0] == bbox1_desc->dims[0]. But now ious_desc->dims[0] is 100, bbox1_desc->dims[0] is 0.<br>4.如果aligned等于1, bbox1_desc和bbox2_desc的第一维必须相等<br>[2023-1-10 17:28:48] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: If it is aligned mode, bbox1_desc->dims[0] == bbox2_desc->dims[0]. But now bbox1_desc->dims[0] is 100, bbox2_desc->dims[0] is 10000.<br>5.如果aligned等于1, bbox1_desc和bbox2_desc的第一维的乘积为0，那么iou_desc的第一维必须是0，第二维必须是1.<br>[2023-1-10 17:32:7] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: If it is aligned mode and rows * cols = 0, ious_desc's first dim should be 0, and ious_desc's last dim should be 1. But now ious_desc's first dim is 0, and ious_desc's last dim is 10000.<br>6.如果aligned等与0,那么iou_desc的dim必须等于2<br>[2023-1-10 17:33:59] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: If it is non-aligned mode, ious_desc->dim == 2. But now ious_desc->dim is 1.<br>7.如果aligned等与0，那么iou_desc的第一维必须和box1_desc的第一维相等，iou_desc的第二维必须和box2_desc的第二维相等<br>[2023-1-10 17:39:1] [MLUOP] [Error]:[mluOpBboxOverlaps] Check failed: If it is non-aligned mode, ious_desc's first dim should be 100, ious_desc's last dim should be 10000.But now ious_desc's first dim is 100, and ious_desc's last dim is 3. |
| 非法参数传递   | 正常报错 |                                                              |

# 2. 性能测试

详见：[MLU-OPS性能验收标准](./MLU-OPS-Performance-Acceptance-Standard.md)

| case  | bbox1_shape | bbox2_shape | iou_shape | mode | aligned | offset | data_type |
| ----- | ----------- | ----------- | --------- | ---- | ------- | ------ | --------- |
| case0 | [240,4]     | [440,4]     | [240,440] | 1    | 0       | 0      | float16   |
| case1 | [880,4]     | [880,4]     | [880,880] | 1    | 1       | 0      | float16   |
| case2 | [240,4]     | [440,4]     | [880,880] | 1    | 0       | 0      | float32   |
| case3 | [880,4]     | [880,4]     | [880,880] | 1    | 1       | 0      | float32   |

平台：MLU370

| case  | mlu_hardware_time(us) | mlu_interface_time(us) | mlu_io_efficiency | mlu_compute_efficiency | mlu_workwpace_size(Bytes) |
| ----- | --------------------- | ---------------------- | ----------------- | ---------------------- | ------------------------- |
| case0 | 587                   | 32.64                  | 0.00120138        | 0.0015541              | 0                         |
| case1 | 35                    | 10.67                  | 0.00147321        | 0.000217205            | 0                         |
| case2 | 584                   | 11.44                  | 0.0024151         | 0.00312418             | 0                         |
| case3 | 49                    | 7.64                   | 0.00210459        | 0.000310292            | 0                         |

# 3. 总结分析

总结分析主要需要考虑以下几点：

1. 此算子不支持含nan和对位都是inf的输入；
2. 综合满足算子验收标准；

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
                    const mluOpTensorDescriptor_t ious_desc,
                    void *ious);
 
+// Group:BboxOverlaps
+/*!
+ * @brief Computes the IOUs or IOFs between two sets of
+ * bounding-boxes. If \b aligned is False, then calculates the IOU of each row between each bounding-box
+ * of \b bbox1 and \b bbox2, otherwise calculates the IOU of the corresponding row between each aligned
+ * pair of \b bbox1 and \b bbox2. For input placed in the order of <x1, y1, x2, y2>, (x1, y1) and (x2, y2)
+ * respectively represents the top-left and bottom-right corner coordinates of bounding-box.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3403]: IOU 和IOF都用单数

represent

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * bounding-box overlaps operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] mode
+ * An integer value which decides to return a result IOUs or IOFs.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3409]: IOU和IOF用单数

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * The integer 0 represents IOUs and 1 represents IOFs.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3410]: IOU和IOF使用单数

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] aligned
+ * A boolean value. If it is False, then calculate the IOUs[i][j] or IOFs[i][j] between
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3412]: false

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * the row i of \b bbox1 and the row j of \b bbox2, otherwise the IOU[i] or IOF[i] between
+ * the row i of \b bbox1 and the row i of \b bbox2 are calculated. The number of row of \b bbox1
+ * and \b bbox2 must be equal for aligned is True.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3415]: for aligned is True  > if \b aligned is true.

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] bbox1
+ * Pointer to the MLU memory that stores the input tensor \b bbox1.
+ * @param[in] bbox2_desc
+ * The descriptor of the input tensor \b bbox2. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] bbox2
+ * Pointer to the MLU memory that stores the input tensor \b bbox2.
+ * @param[in] ious_desc
+ * The descriptor of the output tensor. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[out] ious
+ * IOUs or IOFs. Pointer to the MLU memory that stores the output tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3430]: 单数

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par Formula
+ * - See "Bounding-Box Overlaps Operator" section in "Cambricon BANGC OPS User Guide" for details.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3435]: Operator 》 Operation

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * By the order of \b bbox1 - \b bbox2 - \b ious, the supported data types of
+ * \b bbox1, \b bbox2 and \b ious are as follows:
+ * - float - float - float
+ * - half  - half  - half
+ *
+ * @par Scale Limitation
+ * - The number of dimensions of \b bbox1 and \b bbox2 tensors must be 2.
+ * - The length of lowest dimension of input tensor must be 4.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3445]: the lowest

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *   \b bbox1 (Tensor): shape [m, 4] in <x1, y1, x2, y2> format.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3446]: 去掉句尾的句号，下同

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *   \b bbox2 (Tensor): shape [n, 4] in <x1, y1, x2, y2> format.
+ * - Input with NaN is not supported currently. Also you need to exclude the input with (inf - inf) or (inf - (-inf)),
+ *   where inf represents infinity (because the result is NaN, the actual impact is that the input has NaN).
+ * - For input in type <x1, y1, x2, y2>, the coordinates must satisfy x2 > x1, y2 > y1,
+ *   otherwise incorrect results will be obtained.
+ * - When aligned mode is True, for input \b bbox1 and \b bbox2 with n-rows, if n is zero, the output IOUs
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3452]: True > true

 False > false

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *   array with n-elements. When aligned mode is False, for input \b bbox1 with n-rows and
+ *   \b bbox2 with m-rows, the output IOUs must be a two-dimensional matrix with shape n * m.
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @note
+ * - The input tensor \b x should be in the following range to guarantee the accuracy of output:
+ *  - If bbox_overlaps works on (m)tp_2xx :
+ *    - half : [-300, 100].
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3463]: 句尾句号去掉，下同

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *   must be a two-dimensional matrix with shape n * 1, otherwise the output IOUs must be a one-dimensional
+ *   array with n-elements. When aligned mode is False, for input \b bbox1 with n-rows and
+ *   \b bbox2 with m-rows, the output IOUs must be a two-dimensional matrix with shape n * m.
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @note
+ * - The input tensor \b x should be in the following range to guarantee the accuracy of output:
+ *  - If bbox_overlaps works on (m)tp_2xx :
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3462]: If bbox_overlaps works on (m)tp_2xx：

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -3394,6 +3394,120 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * - The number of dimensions of \b bbox1 and \b bbox2 tensors must be 2.
+ * - The length of lowest dimension of input tensor must be 4.
+ *   \b bbox1 (Tensor): shape [m, 4] in <x1, y1, x2, y2> format.
+ *   \b bbox2 (Tensor): shape [n, 4] in <x1, y1, x2, y2> format.
+ * - Input with NaN is not supported currently. Also you need to exclude the input with (inf - inf) or (inf - (-inf)),
+ *   where inf represents infinity (because the result is NaN, the actual impact is that the input has NaN).
+ * - For input in type <x1, y1, x2, y2>, the coordinates must satisfy x2 > x1, y2 > y1,
+ *   otherwise incorrect results will be obtained.
+ * - When aligned mode is True, for input \b bbox1 and \b bbox2 with n-rows, if n is zero, the output IOUs
+ *   must be a two-dimensional matrix with shape n * 1, otherwise the output IOUs must be a one-dimensional
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3453]: two-dimensional 》 2D

**Review by GraceGuanH — COMMENTED** [review]
Please check it.

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3403]: 已修改

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3409]: 已修改

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3410]: 已修改

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3412]: 已修改

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3415]: 已修改

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3430]: 已修改

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3435]: 已修改

File: bangc-ops/mlu_op.h (hunk 8 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3445]: 已修改

File: bangc-ops/mlu_op.h (hunk 9 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3446]: 已修改

File: bangc-ops/mlu_op.h (hunk 10 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3452]: 已修改

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3463]: 已修改

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3462]: 已修改

File: bangc-ops/mlu_op.h (hunk 13 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3453]: 已修改

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (mahxn0) [reply bangc-ops/mlu_op.h:3403]: Computes the IOUs or IOFs between two sets of
 * bounding-boxes.　此处是两组box计算iou,考虑用复数，保持不变．

**Review by GraceGuanH — APPROVED** [review]

File: bangc-ops/kernels/bbox_overlaps/bbox_overlaps.cpp [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+  if (need_core_num < core_num) {
+    k_dim->y = need_core_num / core_dim;
+  } else {
+    k_dim->y = union_num;
+  }
+  k_dim->z = 1;
+  return;
+}
+
+mluOpStatus_t mluOpBboxOverlaps(
```
Comment (wushaoqiang2168) [inline bangc-ops/kernels/bbox_overlaps/bbox_overlaps.cpp:51]: 缺少MLUOP_WIN_API

Comment (mahxn0) [reply bangc-ops/kernels/bbox_overlaps/bbox_overlaps.cpp:51]: 已修改

**Review by DanieeelLiu — APPROVED** [review]

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**